bearer/bearer.go

```diff
@@ -2,335 +2,330 @@ package bearer
+// AssertUser checks if the Token is issued to the given user.
+//
+// Zero Token is available to any user.
+//
+// See also ForUser.
+func (b Token) AssertUser(id user.ID) bool {
+	return !b.targetUserSet || b.targetUser.Equals(id)
+}
 
-	if !sig.Verify(m.GetBody().StableMarshal(nil)) {
-		return errors.New("wrong signature")
+// Sign calculates and writes signature of the Token data using issuer's secret.
+// Returns signature calculation errors.
+//
+// Sign MUST be called if Token is going to be transmitted over
+// NeoFS API V2 protocol.
+//
+// Note that any Token mutation is likely to break the signature, so it is
+// expected to be calculated as a final stage of Token formation.
+//
+// See also VerifySignature, Issuer.
+func (b *Token) Sign(key ecdsa.PrivateKey) error {
+	err := b.sig.Calculate(neofsecdsa.Signer(key), b.signedData())
+	if err == nil {
+		b.sigSet = true
 	}
 
 	return nil
 }
 
-// Issuer returns user.ID associated with the key that signed bearer token.
-// To pass node validation it should be owner of requested container.
+// VerifySignature checks if Token signature is presented and valid.
 //
-// If token is not signed, Issuer returns empty owner ID and false `ok` flag.
+// Zero Token fails the check.
 //
 // See also Sign.
-func (b Token) Issuer() (id user.ID, ok bool) {
-	v2 := (acl.BearerToken)(b)
-
-	pub, _ := keys.NewPublicKeyFromBytes(v2.GetSignature().GetKey(), elliptic.P256())
-
-	ok = pub != nil
-	if ok {
-		user.IDFromKey(&id, (ecdsa.PublicKey)(*pub))
-	}
-
-	return
+func (b Token) VerifySignature() bool {
+	// TODO: (#233) check owner<->key relation
+	return b.sigSet && b.sig.Verify(b.signedData())
 }
 
-// sanityCheck if bearer token is ready to be issued.
-func sanityCheck(b *Token) error {
-	v2 := (*acl.BearerToken)(b)
-
-	switch {
-	case v2.GetBody() == nil:
-		return errNilBearerTokenBody
-	case v2.GetBody().GetEACL() == nil:
-		return errNilBearerTokenEACL
-	}
-
-	// consider checking ContainerEACL sanity there, lifetime correctness, etc.
-
-	return nil
-}
-
-// Marshal marshals Token into a canonical NeoFS binary format (proto3
-// with direct field order).
+// Marshal encodes Token into a binary format of the NeoFS API protocol
+// (Protocol Buffers V3 with direct field order).
 //
 // See also Unmarshal.
 func (b Token) Marshal() []byte {
-	return (*acl.BearerToken)(&b).StableMarshal(nil)
+	var m acl.BearerToken
+	b.WriteToV2(&m)
+
+	return m.StableMarshal(nil)
 }
 
-// Unmarshal unmarshals Token from canonical NeoFS binary format (proto3
-// with direct field order).
+// Unmarshal decodes NeoFS API protocol binary data into the Token
+// (Protocol Buffers V3 with direct field order). Returns an error describing
+// a format violation.
 //
 // See also Marshal.
 func (b *Token) Unmarshal(data []byte) error {
-	v2 := (*acl.BearerToken)(b)
-	return v2.Unmarshal(data)
+	var m acl.BearerToken
+
+	err := m.Unmarshal(data)
+	if err != nil {
+		return err
+	}
+
+	return b.readFromV2(m, false)
```

Comment:
why dont we check the format on unmarshalling? what is the difference b/w constructing SDK struct from v2 (that could probably be contracted from raw invalid bytes) and constructing from raw bytes?

Reply:
What do u mean incorrect struct? Is unsigned struct incorrect? It is completely valid in binary and JSON representations. I'd say "incorrect" only about the messages - which violate the protocol. But `Token` isn't a message, it can be read from and written into message.